guppylang/compiler/core.py

```diff
@@ -45,28 +21,49 @@ class DFContainer:
     current compilation state.
     """
 
+    graph: Hugr
     node: DFContainingNode
-    locals: CompiledLocals
-
-    def __getitem__(self, item: str) -> PortVariable:
-        return self.locals[item]
-
-    def __setitem__(self, key: str, value: PortVariable) -> None:
-        self.locals[key] = value
-
-    def __iter__(self) -> Iterator[PortVariable]:
-        return iter(self.locals.values())
-
-    def __contains__(self, item: str) -> bool:
-        return item in self.locals
+    locals: CompiledLocals = field(default_factory=dict)
+
+    def __getitem__(self, place: Place) -> OutPortV:
+        """Constructs a port for a local place in this DFG.
+
+        Note that this mutates the Hugr since we might need to pack or unpack some
+        tuples to obtain a port for places that involve struct fields.
+        """
+        # First check, if we already have a wire for this place
+        if place.id in self.locals:
+            return self.locals[place.id]
+        # Otherwise, our only hope is that it's a struct value that we can rebuild by
+        # packing the wires of its constituting fields
+        if not isinstance(place.ty, StructType):
+            raise InternalGuppyError(f"Couldn't obtain a port for `{place}`")
+        children = [FieldAccess(place, field, None) for field in place.ty.fields]
+        child_ports = [self[child] for child in children]
+        port = self.graph.add_make_tuple(child_ports, self.node).out_port(0)
+        for child in children:
+            if child.ty.linear:
+                self.locals.pop(child.id)
+        self.locals[place.id] = port
+        return port
+
+    def __setitem__(self, place: Place, port: OutPortV) -> None:
+        # When assigning a struct value, we immediately unpack it recursively and only
+        # store the leaf wires.
+        is_return = isinstance(place, Variable) and is_return_var(place.name)
+        if isinstance(place.ty, StructType) and not is_return:
+            unpack = self.graph.add_unpack_tuple(port, self.node)
+            for field, field_port in zip(
+                place.ty.fields, unpack.out_ports, strict=False
+            ):
+                self[FieldAccess(place, field, None)] = field_port
```

Comment:
I think we should delete `self.locals[place.id]` here?
Otherwise reading the whole struct would still use the old value.

Reply:
Good catch! Here is a minimal example that currently fails validation because of this:

```python
@guppy.struct(module)
class MyStruct:
    x: int


@guppy(module)
def foo() -> MyStruct:
    s = 0
    s = MyStruct(42)
    return s  # Wires up the 0 value since `s = 0` is still in scope
```

I'll fix it and add a test

Reply:
See b7bc393 for the test